s/common_test.go

```diff
@@ -0,0 +1,359 @@
+package s_test
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/capatazlib/go-capataz/c"
+	"github.com/capatazlib/go-capataz/s"
+)
+
+////////////////////////////////////////////////////////////////////////////////
+
+// EventP is a predicate like function that allows us to assert properties of an
+// Event signaled by the supervision system
+type EventP interface {
+	// Call executes the EventP predicate logic
+	Call(s.Event) bool
+
+	// Returns an string representation of the Predicate (for debugging purposes)
+	String() string
+}
+
+// EventTagP is a predicate to assert the EventTag of a given event matches
+// the expected EventTag
+type EventTagP struct {
+	tag s.EventTag
+}
+
+func (p EventTagP) Call(ev s.Event) bool {
+	return ev.Tag() == p.tag
+}
+
+func (p EventTagP) String() string {
+	return fmt.Sprintf("tag == %s", p.tag.String())
+}
+
+// ProcessNameP is a predicate to assert the process name of a given event
+// matches the expected name
+type ProcessNameP struct {
+	name string
+}
+
+func (p ProcessNameP) Call(ev s.Event) bool {
+	return ev.Name() == p.name
+}
+
+func (p ProcessNameP) String() string {
+	return fmt.Sprintf("name == %s", p.name)
+}
+
+// AndP is a predicate that does of a conjunction of many EventP predicates
+type AndP struct {
+	preds []EventP
+}
+
+func (p AndP) Call(ev s.Event) bool {
+	acc := true
+	for _, pred := range p.preds {
+		acc = acc && pred.Call(ev)
+		if !acc {
+			return acc
+		}
+	}
+	return acc
+}
+
+func (p AndP) String() string {
+	acc := make([]string, 0, len(p.preds))
+	for _, pred := range p.preds {
+		acc = append(acc, pred.String())
+	}
+	return strings.Join(acc, " && ")
+}
+
+// ProcessStarted is a predicate to assert an event represents a process that
+// got started
+func ProcessStarted(name string) EventP {
+	return AndP{
+		preds: []EventP{
+			EventTagP{tag: s.ProcessStarted},
+			ProcessNameP{name: name},
+		},
+	}
+}
+
+// ProcessStopped is a predicate to assert an event represents a process that
+// got stopped by its parent supervisor
+func ProcessStopped(name string) EventP {
+	return AndP{
+		preds: []EventP{
+			EventTagP{tag: s.ProcessStopped},
+			ProcessNameP{name: name},
+		},
+	}
+}
+
+// Leaving this for later
+// func ProcessFailed(name string) EventP {
+//	return AndP{
+//		preds: []EventP{
+//			EventTagP{tag:s.ProcessFailed},
+//			ProcessNameP{name:name},
+//		}
+//	}
+// }
+
+////////////////////////////////////////////////////////////////////////////////
+
+// noWait is a notification callback function that never waits a condition
+func noWait() {}
+
+// startEventCollector collects all the events sent to the given channel and
+// returns a slice with all the events captured till the channel was closed
+func startEventCollector(evCh chan s.Event) func() []s.Event {
+	buffer := make([]s.Event, 0, 10)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
```

Comment:
I would a couple of notes indicating that this goro collects the event, while they come, while the returned function will close the event channel at a later time. It took me a couple of reads to figure out what was going on.

Reply:
Added a few comments to the code, let me know if they help